### PR TITLE
fix(mutagen): use shorter directories for mutagen syncs

### LINKED ISSUE #4527: 0.13: [Bug]: in-cluster build sync error when root path is very long
### Current Behavior

When using `garden build` in especially long root paths (over 70 chars) I've noticed an odd repeating error (log below).

For example, the full path being used below is `/tmp/Extremely/Long/path-with/some-hyphens/and/whatnot/example-bonsai-sync-issue`

<details>
<summary>command output</summary>

```
❯ KUBE_CONTEXT=amer-dev REGISTRY_HOSTNAME=reg-chrisprzybycien.tools.splashdevelop.com garden build
Build 🔨

Garden v0.13 (Bonsai) is a major release with significant changes. Please help us improve it by reporting any issues/bugs here:
https://go.garden.io/report-bonsai
→ Run garden util hide-warning 0.13-bonsai to disable this warning.
ℹ garden               → Running in Garden environment example.chrisprzybycien
ℹ providers            → Getting status...
✔ providers            → Cached (took 0.6 sec)
ℹ providers            → Run with --force-refresh to force a refresh of provider statuses.
ℹ graph                → Resolving actions and modules...
✔ graph                → Done (took 0.1 sec)
ℹ build.base-image     → -> Deploying garden-buildkit daemon in example-problem-chrisprzybycien namespace (was outdated)
ℹ build.base-image     → Waiting for resources to be ready...
ℹ build.base-image     → Resources ready
ℹ build.base-image     → Done!
ℹ build.base-image     → -> Deploying garden-buildkit daemon in example-problem-chrisprzybycien namespace (was outdated)
ℹ build.base-image     → Waiting for resources to be ready...
ℹ build.base-image     → Resources ready
ℹ build.base-image     → Done!
⚠ build.base-image     → Synchronization monitor exited with code 1.
⚠ build.base-image     → Could not connect to sync daemon, retrying (attempt 1/10)...
⚠ build.base-image     → Could not connect to sync daemon, retrying (attempt 2/10)...
⚠ build.base-image     → Synchronization monitor exited with code 1.
⚠ build.base-image     → Could not connect to sync daemon, retrying (attempt 3/10)...
⚠ build.base-image     → Synchronization monitor exited with code 1.
⚠ build.base-image     → Synchronization monitor exited with code 1.
⚠ build.base-image     → Could not connect to sync daemon, retrying (attempt 4/10)...
⚠ build.base-image     → Synchronization monitor exited with code 1.
⚠ build.base-image     → Could not connect to sync daemon, retrying (attempt 5/10)...
⚠ build.base-image     → Synchronization monitor exited with code 1.
⚠ build.base-image     → Could not connect to sync daemon, retrying (attempt 6/10)...
⚠ build.base-image     → Synchronization monitor exited with code 1.
⚠ build.base-image     → Could not connect to sync daemon, retrying (attempt 7/10)...
⚠ build.base-image     → Synchronization monitor exited with code 1.
⚠ build.base-image     → Could not connect to sync daemon, retrying (attempt 8/10)...
⚠ build.base-image     → Synchronization monitor exited with code 1.
⚠ build.base-image     → Synchronization monitor exited with code 1.
⚠ build.base-image     → Could not connect to sync daemon, retrying (attempt 9/10)...
⚠ build.base-image     → Synchronization monitor exited with code 1.
⚠ build.base-image     → Could not connect to sync daemon, retrying (attempt 10/10)...
⚠ build.base-image     → Synchronization monitor exited with code 1.
⚠ build.base-image     → Synchronization monitor exited with code 1.
⚠ build.base-image     → Could not connect to sync daemon, retrying (attempt 1/10)...
⚠ build.base-image     → Synchronization monitor exited with code 1.
⚠ build.base-image     → Could not connect to sync daemon, retrying (attempt 2/10)...
⚠ build.base-image     → Synchronization monitor exited with code 1.
⚠ build.base-image     → Could not connect to sync daemon, retrying (attempt 3/10)...
⚠ build.base-image     → Synchronization monitor exited with code 1.
⚠ build.base-image     → Could not connect to sync daemon, retrying (attempt 4/10)...
⚠ build.base-image     → Synchronization monitor exited with code 1.
⚠ build.base-image     → Could not connect to sync daemon, retrying (attempt 5/10)...
⚠ build.base-image     → Synchronization monitor exited with code 1.
⚠ build.base-image     → Could not connect to sync daemon, retrying (attempt 6/10)...
⚠ build.base-image     → Synchronization monitor exited with code 1.
⚠ build.base-image     → Synchronization monitor exited with code 1.
⚠ build.base-image     → Could not connect to sync daemon, retrying (attempt 7/10)...
⚠ build.base-image     → Synchronization monitor exited with code 1.
⚠ build.base-image     → Could not connect to sync daemon, retrying (attempt 8/10)...
⚠ build.base-image     → Synchronization monitor exited with code 1.
⚠ build.base-image     → Could not connect to sync daemon, retrying (attempt 9/10)...
⚠ build.base-image     → Synchronization monitor exited with code 1.
⚠ build.base-image     → Synchronization monitor exited with code 1.
⚠ build.base-image     → Could not connect to sync daemon, retrying (attempt 10/10)...
⚠ build.base-image     → Synchronization monitor exited with code 1.
✖ build.base-image     → Build failed (took 248.2 sec)
✖ build.base-image     → Failed processing Build type=container name=base-image (took 248.22 sec). Here is the output:

────────────────────────────────────────────────────────────────────────────────
Command "/Users/chrisprzybycien/.garden/tools/mutagen/5f82909105ed5d86/mutagen sync terminate k8s--build-sync--example--example-problem-chrisprzybycien--base-image--bubrjv4c" failed with code 1:

Attempting to start Mutagen daemon...
Error: unable to connect to daemon: connection timed out (is the daemon running?)
────────────────────────────────────────────────────────────────────────────────
1 build action(s) failed!

See .garden/error.log for detailed error message
❯ pwd
/tmp/Extremely/Long/path-with/some-hyphens/and/whatnot/example-bonsai-sync-issue

 /tmp/Extremely/Long/path-with/some-hyphens/and/whatnot/example-bonsai-sync-issue  main ?1 
❯
```
</details>

### Expected behavior

Garden builds the container without issue.

### Reproducible example

I created a super simple example in a repo called [example-bonsai-sync-issue](https://github.com/chrispsplash/example-bonsai-sync-issue)

### Workaround

There is no problem when you move the project to a much shorter path (ex `/tmp/example-bonsai-sync-issue`).

### Suggested solution(s)

🤷 

### Additional context

N/A

### Your environment

* **OS:** MacOS Venture 13.2.1
* **How I'm running Kubernetes:** EKS

`garden version`
0.13.0

### PR DESCRIPTION
**Pull request by shumailxyz**
**What this PR does / why we need it**:
Use directory at shorter path for mutagen syncs if the project path is longer than `MUTAGEN_DATA_DIRECTORY_LENGTH_LIMIT`. The shorter directory path is computed using hash of the path of .garden dir in project and appending to `$HOME/.garden/mutagen`.

**Which issue(s) this PR fixes**:
Fixes #4527 

**Special notes for your reviewer**:

### CODE REVIEW

File: core/src/mutagen.ts [hunk 1]
```diff
@@ -331,7 +332,7 @@ export class Mutagen {
   constructor({ ctx, log, dataDir }: MutagenDaemonParams) {
     this.log = log
     this.configLock = new AsyncLock()
-    this.dataDir = dataDir || join(ctx.gardenDirPath, MUTAGEN_DIR_NAME)
+    this.dataDir = dataDir || getMutagenDataDir(ctx.gardenDirPath, log)
```
Comment (vvagaytsev) [inline core/src/mutagen.ts:335]: I've just checked the code, and it looks like we never provide a value of the optional `dataDir?` in the `MutagenDaemonParams`.

I'm just wondering if we should remove it from the parameter object and always infer it from the context.

File: core/src/mutagen.ts [hunk 2]
```diff
@@ -778,13 +779,29 @@ export interface SyncSession {
...
+/**
+ * Returns mutagen data directory path based on the project dir.
+ * If the project path longer than `MUTAGEN_DATA_DIRECTORY_LENGTH_LIMIT`, it computes
+ * hash of project dir path, uses first 9 characters of hash as directory name
+ * and creates a directory in $HOME/.garden/mutagen.
+ *
+ * However, if the path is not longer than `MUTAGEN_DATA_DIRECTORY_LENGTH_LIMIT`, then
+ * it uses the ./project-root/.garden/mutagen directory.
+ */
+export function getMutagenDataDir(path: string, log: Log) {
```
Comment (vvagaytsev) [inline core/src/mutagen.ts:791]: This looks good! 💯 
Let's give it an excessive manual testing round to make sure everything works as expected.

By the way, do we already have the machinery to cleanup the temp sync directories? I bet we do, but let's double-check.

Comment (shumailxyz) [reply core/src/mutagen.ts:791]: I doubt if we were cleaning up the mutagen temp directory as it was located in `.garden` dir in project and i couldn't find it in code if we are cleaning up that dir or mutagen dir inside.

For tests, we area cleaning up already.

Just wondering, when should we even cleanup the mutagen directory? As long as the project exists, mutagen directory should also exist in my opinion.

Comment (vvagaytsev) [reply core/src/mutagen.ts:791]: Ah, right, sure :) I just anted to be sure that we're not missing anything :)

**Comment by shumailxyz** [comment]
@vvagaytsev @Walther Ready from my side now. Please give it a spin.

**Review by Walther — APPROVED** [review]
Thank you for the fix! ✨ 
Verified working locally, no more long path warnings 🧹

**Comment by shumailxyz** [comment]
@Walther For the warning, i basically removed the previous warning completely and added a verbose level log statement: https://github.com/garden-io/garden/pull/4867/files#diff-76ea7f738d0aaad52a9e3339c18c153870b2229d85b5efb7fb3c27223dc96aa3R795

**Review by vvagaytsev — APPROVED** [review]
Great job, thanks! 💯

**Pull request merged**